src/Microsoft.ML.Data/Transforms/ConvertTransform.cs

```diff
@@ -122,298 +145,194 @@ public class Arguments : TransformInputBase
         internal const string LoaderSignatureOld = "ConvertFunction";
         private static VersionInfo GetVersionInfo()
         {
             return new VersionInfo(
                 modelSignature: "CONVERTF",
                 // verWrittenCur: 0x00010001, // Initial
-                verWrittenCur: 0x00010002, // Added support for keyRange
-                verReadableCur: 0x00010002,
-                verWeCanReadBack: 0x00010001,
+                //verWrittenCur: 0x00010002, // Added support for keyRange
+                verWrittenCur: 0x00010003, // Change to transformer leads to change of saving objects.
+                verReadableCur: 0x00010003,
+                verWeCanReadBack: 0x00010003,
                 loaderSignature: LoaderSignature,
                 loaderSignatureAlt: LoaderSignatureOld,
                 loaderAssemblyName: typeof(ConvertTransform).Assembly.FullName);
         }
 
         private const string RegistrationName = "Convert";
 
-        // This is parallel to Infos.
-        private readonly ColInfoEx[] _exes;
-
+        public IReadOnlyCollection<ColumnInfo> Columns => _columns.AsReadOnly();
         /// <summary>
-        /// Convenience constructor for public facing API.
+        /// Describes how the transformer handles one column pair.
         /// </summary>
-        /// <param name="env">Host Environment.</param>
-        /// <param name="input">Input <see cref="IDataView"/>. This is the output from previous transform or loader.</param>
-        /// <param name="resultType">The expected type of the converted column.</param>
-        /// <param name="name">Name of the output column.</param>
-        /// <param name="source">Name of the column to be converted.  If this is null '<paramref name="name"/>' will be used.</param>
-        public ConvertTransform(IHostEnvironment env,
-            IDataView input,
-            DataKind resultType,
-            string name,
-            string source = null)
-            : this(env, new Arguments() { Column = new[] { new Column() { Source = source ?? name, Name = name } }, ResultType = resultType }, input)
-        {
-        }
-
-        public ConvertTransform(IHostEnvironment env, Arguments args, IDataView input)
-            : base(env, RegistrationName, env.CheckRef(args, nameof(args)).Column,
-                input, null)
+        public sealed class ColumnInfo
         {
-            Host.AssertNonEmpty(Infos);
-            Host.Assert(Infos.Length == Utils.Size(args.Column));
+            public readonly string Input;
+            public readonly string Output;
+            public readonly DataKind Kind;
+            public readonly KeyRange KeyRange;
 
-            _exes = new ColInfoEx[Infos.Length];
-            for (int i = 0; i < _exes.Length; i++)
+            public ColumnInfo(string input, string output, DataKind kind, KeyRange keyRange = null)
             {
-                DataKind kind;
-                KeyRange range;
```

Comment:
>KeyRange [](start = 37, length = 8)

is it at all possible to fold `Kind` and `KeyRange` into one `ColumnType ItemType`?..

Reply:
maybe but I would prefer to concentrate on Estimator/Transformer conversion rather than optimizing how it works.

---
In reply to: [229429947](https://github.com/dotnet/machinelearning/pull/1439#discussion_r229429947) [](ancestors = 229429947)